Check gas limit consistency with the target (#5236)
ethereum/consensus-specs: specs/gloas/p2p-interface.md

`signed_proposer_preferences` for the validated `SignedProposerPreferences`
whose `message.proposal_slot` is `bid.slot` and `message.dependent_root` is
`get_proposer_dependent_root(parent_state, compute_epoch_at_slot(bid.slot))`,

> where `parent_block` is the block with block root equal to

Reviewer: `parent_block` is not used

Author: yeah, this is because I wanted to use it in the helper but decided to leave it anyway cause post state of a root is not something I liked, feel free to remove it if you don't like how it reads now

Reviewer: can keep it too, no strong opinion, leaving up to @jtraglia

Reviewer 2: I'd suggest keeping it. It'll change soon anyway.

Reviewer: I kept the change from @potuz, merged my other changes from https://github.com/potuz/consensus-specs/pull/21 into this PR

- _[IGNORE]_ `bid.value` is less or equal than the builder's excess balance --
  i.e. `can_builder_cover_bid(state, builder_index, amount)` returns `True`.
- _[IGNORE]_ `bid.parent_block_hash` is the block hash of a known execution
  payload in fork choice and

>   `is_gas_limit_target_compatible(parent_gas_limit, bid.gas_limit, proposer_preferences.target_gas_limit)`

Reviewer: just realized this but shouldn't this be a `REJECT` if `is_gas_limit_target_compatible = False`? we do the same if fee recipient is incorrect, wondering if we used `IGNORE` just because it's combined with "known execution payload in fork choice" condition

Reviewer 2: Yes, I think it should. Can you open a PR for this?

Reviewer: - https://github.com/ethereum/consensus-specs/pull/5428

  is `True` where `parent_gas_limit` is the `gas_limit` of that execution
  payload.
- _[IGNORE]_ `bid.parent_block_root` is the hash tree root of a known beacon
  block in fork choice.
- _[REJECT]_ `signed_execution_payload_bid.signature` is valid with respect to
  the `bid.builder_index`.

```python
def is_gas_limit_target_compatible(
    parent_gas_limit: uint64, gas_limit: uint64, target_gas_limit: uint64
) -> bool:
    """
    Check if ``gas_limit`` is compatible with ``target_gas_limit`` under the
    EIP-1559 transition rule from ``parent_gas_limit``.
    """
    max_gas_limit_difference = max(parent_gas_limit // 1024, 1) - 1

>     min_gas_limit = parent_gas_limit - max_gas_limit_difference

Reviewer 3: Don't we have to worry about overflow here? What if `parent_gas_limit` is really small (or 0, as we encountered at default Gloas genesis).

Reviewer 3: Actually nvm, it's fine because of how `max_gas_limit_difference` is calculated (`parent_gas_limit <= 1024` implies `max_gas_limit_difference == 0`).

- _[IGNORE]_ `bid.value` is less or equal than the builder's excess balance --
  i.e. `can_builder_cover_bid(state, builder_index, amount)` returns `True`.
- _[IGNORE]_ `bid.parent_block_hash` is the block hash of a known execution
  payload in fork choice. Let `parent_bid` be the `ExecutionPayloadBid` in the

>   beacon block such that `parent_bid.block_hash == bid.parent_block_hash`.

Reviewer: this does not work at the fork boundary

- _[IGNORE]_
  `is_gas_limit_target_compatible(parent_bid.gas_limit, bid.gas_limit, proposer_preferences.gas_limit) == True`.
- _[IGNORE]_ `bid.parent_block_root` is the hash tree root of a known beacon
  block in fork choice.
- _[REJECT]_ `signed_execution_payload_bid.signature` is valid with respect to
  the `bid.builder_index`.

The following helper is used to validate the bid's gas limit

```python
def is_gas_limit_target_compatible(
    parent_gas_limit: uint64, gas_limit: uint64, proposer_gas_limit: uint64
) -> bool:
    """

>     Check if the bid's gas limit is compatible with the proposer's preferences.

Reviewer: could add more details here

```suggestion
    Check if ``gas_limit`` is compatible with ``proposer_gas_limit`` under the
    EIP-1559 transition rule from ``parent_gas_limit``.
```

    """

>     min_gas_limit = parent_gas_limit - parent_gas_limit // 1024

Reviewer 4: This helper is off by one relative to execution-layer gas-limit validity.

The EL check rejects both boundary values:

```python
max_adjustment_delta = parent_gas_limit // GasCosts.LIMIT_ADJUSTMENT_FACTOR
if gas_limit >= parent_gas_limit + max_adjustment_delta:
    return False
if gas_limit <= parent_gas_limit - max_adjustment_delta:
    return False
```

Source: https://github.com/ethereum/execution-specs/blob/44923d4bcdbcaa48ef1536a453c390a0ebf08220/src/ethereum/forks/amsterdam/fork.py#L1147-L1155

So the valid interval is strict:

```python
parent_gas_limit - delta < gas_limit < parent_gas_limit + delta
```

But this helper currently sets:

```python
min_gas_limit = parent_gas_limit - delta
max_gas_limit = parent_gas_limit + delta
```

and can accept/require those EL-invalid boundary values.

For example:

```python
parent_gas_limit = 30_000_000
delta = parent_gas_limit // 1024  # 29_296

# EL-valid maximum is 30_029_295
# Current helper clamps to 30_029_296, which EL rejects
```

A small repro against this helper is:

```python
parent_gas_limit = spec.uint64(30_000_000)
delta = parent_gas_limit // 1024
proposer_gas_limit = spec.uint64(parent_gas_limit + delta + 1)
highest_el_valid_gas_limit = spec.uint64(parent_gas_limit + delta - 1)

assert spec.is_gas_limit_target_compatible(
    parent_gas_limit,
    highest_el_valid_gas_limit,
    proposer_gas_limit,
)
```

This currently fails because the helper expects `parent + delta`, but the nearest EL-valid value is `parent + delta - 1`.

I think the bounds should be:

```python
min_gas_limit = parent_gas_limit - parent_gas_limit // 1024 + 1
max_gas_limit = parent_gas_limit + parent_gas_limit // 1024 - 1
```

with the in-range check using those same inclusive CL helper bounds.

Reviewer 2: This is a really good catch. Thanks @kevaundray.

Author: Thanks, fixed in dd31277

Reviewer: added a test for this in https://github.com/ethereum/consensus-specs/pull/5236/changes/e92f06f4d8e9db4d4995f2b78820cebe7d859b3d

>     max_gas_limit = parent_gas_limit + parent_gas_limit // 1024

Reviewer: compared this against lodestar implementation, this is off by one

```suggestion
    max_gas_limit_difference = parent_gas_limit // 1024 - 1
    min_gas_limit = parent_gas_limit - max_gas_limit_difference
    max_gas_limit = parent_gas_limit + max_gas_limit_difference
```

Reviewer: could adopt similar test cases we have [here](https://github.com/ChainSafe/lodestar/blob/unstable/packages/beacon-node/test/unit/execution/builder/utils.test.ts)

Author: Fixed in dd31277

Reviewer: add some tests in https://github.com/ethereum/consensus-specs/pull/5236/changes/e92f06f4d8e9db4d4995f2b78820cebe7d859b3d

> 
> ```python
> def is_gas_limit_target_compatible(
>     parent_gas_limit: uint64, gas_limit: uint64, preferences_gas_limit: uint64
> ) -> bool:
>     """
>     Check if the bid's gas limit is compatible with the proposer's preferences.
>     """
>     min_gas_limit = parent_gas_limit - parent_gas_limit // 1024
>     max_gas_limit = parent_gas_limit + parent_gas_limit // 1024
> 
>     if preferences_gas_limit >= min_gas_limit and preferences_gas_limit <= max_gas_limit:
>         return gas_limit == preferences_gas_limit
>     if preferences_gas_limit > max_gas_limit:
>         return gas_limit == max_gas_limit
>     return gas_limit == min_gas_limit

Reviewer 2: I think it would read slightly better with "proposer" here instead.
```suggestion
def is_gas_limit_target_compatible(
    parent_gas_limit: uint64, gas_limit: uint64, proposer_gas_limit: uint64
) -> bool:
    """
    Check if the bid's gas limit is compatible with the proposer's preferences.
    """
    min_gas_limit = parent_gas_limit - parent_gas_limit // 1024
    max_gas_limit = parent_gas_limit + parent_gas_limit // 1024

    if proposer_gas_limit >= min_gas_limit and proposer_gas_limit <= max_gas_limit:
        return gas_limit == proposer_gas_limit
    if proposer_gas_limit > max_gas_limit:
        return gas_limit == max_gas_limit
    return gas_limit == min_gas_limit
```
